src/modules/variable/get.rs

```diff
@@ -61,29 +76,66 @@ impl SyntaxModule<ParserMetadata> for VariableGet {
 }
 
 impl TranslateModule for VariableGet {
+    #[allow(clippy::collapsible_else_if)]
     fn translate(&self, meta: &mut TranslateMetadata) -> String {
         let name = self.get_translated_name();
-        let ref_prefix = if self.is_ref { "!" } else { "" };
-        let res = format!("${{{ref_prefix}{name}}}");
         // Text variables need to be encapsulated in string literals
         // Otherwise, they will be "spread" into tokens
         let quote = meta.gen_quote();
-        match (self.is_ref, &self.kind) {
-            (false, Type::Array(_)) => match *self.index {
-                Some(ref expr) => format!("{quote}${{{name}[{}]}}{quote}", expr.translate(meta)),
-                None => format!("{quote}${{{name}[@]}}{quote}")
-            },
-            (true, Type::Array(_)) => match *self.index {
-                Some(ref expr) => {
-                    let id = meta.gen_value_id();
-                    let expr = expr.translate_eval(meta, true);
-                    meta.stmt_queue.push_back(format!("eval \"local __AMBER_ARRAY_GET_{id}_{name}=\\\"\\${{${name}[{expr}]}}\\\"\""));
-                    format!("$__AMBER_ARRAY_GET_{id}_{name}") // echo $__ARRAY_GET
-                },
-                None => format!("{quote}${{!__AMBER_ARRAY_{name}}}{quote}")
-            },
-            (_, Type::Text) => format!("{quote}{res}{quote}"),
-            _ => res
+        match &self.kind {
+            Type::Array(_) => {
+                if self.is_ref {
+                    if let Some(index) = self.index.as_ref() {
+                        let value = match &index.value {
+                            Some(ExprType::Range(range)) => {
+                                let (offset, length) = range.get_array_index(meta);
+                                format!("\\\"\\${{${name}[@]:{offset}:{length}}}\\\"")
+                            }
+                            Some(ExprType::Neg(neg)) => {
+                                let index = neg.get_array_index(meta);
+                                format!("\\\"\\${{${name}[{index}]}}\\\"")
+                            }
+                            _ => {
+                                let index = index.translate_eval(meta, true);
+                                format!("\\\"\\${{${name}[{index}]}}\\\"")
+                            }
+                        };
+                        let id = meta.gen_value_id();
+                        let stmt = format!("eval \"local __AMBER_ARRAY_GET_{id}_{name}={value}\"");
+                        meta.stmt_queue.push_back(stmt);
+                        format!("$__AMBER_ARRAY_GET_{id}_{name}") // echo $__ARRAY_GET
```

Comment:
Is it possible to create a separate function for handling the indexed variable?

Reply:
Refactored.